generate/state.go

```diff
@@ -183,23 +193,92 @@ func FromState(tfstate json.RawMessage, opt Options) (*graph.Graph, map[string]i
 	return g, endCfg, nil
 }
 
-// validateTFStateVersion validates that the verion is the
-// one we support which is only 4
+// migrateVersions will try to apply migrations of old
+// statefile:
+// * For version 3 will try to populate the Dependencies as they are
+//   removed by TF as they cannot be migrated from v3->v4
+func migrateVersions(src []byte, f *statefile.File) error {
+	version, err := tfStateVersion(src)
+	if err != nil {
+		return fmt.Errorf("could not read version: %w", err)
+	}
+
+	deps := make(map[string][]addrs.ConfigResource)
+
+	var state map[string]interface{}
+
+	if version == 3 {
+		err := json.Unmarshal(src, &state)
+		if err != nil {
+			return fmt.Errorf("error while reading TFState: %w", err)
+		}
+
+		for _, m := range state["modules"].([]interface{}) {
+			rs, ok := m.(map[string]interface{})["resources"]
+			if !ok {
+				continue
+			}
+			for rn, rv := range rs.(map[string]interface{}) {
+				// As we have replaced from 'depends_on' to 'dependencies'
+				// we have to access it with 'dependencies'
+				dps, ok := rv.(map[string]interface{})["dependencies"]
+				if !ok && dps == nil {
+					continue
+				}
+				for _, d := range dps.([]interface{}) {
+					ds := strings.Split(d.(string), ".")
+					deps[rn] = append(deps[rn], addrs.ConfigResource{
+						Resource: addrs.Resource{
+							Mode: addrs.ManagedResourceMode,
+							Type: ds[0],
+							Name: ds[1],
+						},
+					})
+				}
+			}
+		}
+
+		for _, v := range f.State.Modules {
+			for rk, rv := range v.Resources {
+				for _, iv := range rv.Instances {
+					if d, ok := deps[rk]; ok {
+						iv.Current.Dependencies = d
+					}
+				}
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// validateTFStateVersion validates that the version is the
+// one we support which is only 3 and 4
 func validateTFStateVersion(b []byte) error {
+	v, err := tfStateVersion(b)
+	if err != nil {
+		return fmt.Errorf("could not read version: %w", err)
+	}
+
+	if v != 4 && v != 3 {
+		return fmt.Errorf("could not read version %d: %w", v, errcode.ErrInvalidTFStateVersion)
```

Comment:
Shouldn't we precise like: "only version X & Y are supported" to be more explicit?

Reply:
ok I've added it but to the `ErrInvalidTFStateVersion` as it has already the information for that.